### PR TITLE
Add IPv6 listener support; Make listen options configurable

### PR DESCRIPTION
**Pull request by shokinn**
This pull request adds the possibility to open an IPv6 socket and make the listen options configurable
(besides 'ssl' for ssl enabled servers and 'ipv6only=on' for IPv6 listeners).

**This change will break the current behavior how ports are defined!**

Old way:  
```
port: 8080
```

New way:  
```
listen:
  port: 8080
  opts: # Listen opts like http2 which will be added (ssl is automatically added if you specify 'ssl:').
    - default_server
  ipv4: true
  ipv6: true
```

Ref:
  - #146 
  - #85

### CODE REVIEW

File: templates/http/default.conf.j2 [hunk 1]
```diff
@@ -51,7 +51,12 @@ auth_request {{ item.value.auth_request_http }};
 
 server {
 {% if item.value.ssl is defined and item.value.ssl %}
-    listen {{ item.value.port }} ssl;
+{% if ( item.value.listen.ipv4 is defined and item.value.listen.ipv4 ) or ( item.value.listen.ipv4 is not defined ) %}
+    listen {{ item.value.listen.port }} ssl{% if item.value.listen.opts is defined and item.value.listen.opts|length %} {{ item.value.listen.opts|join(" ") }}{% endif %};
```
Comment (alessfg) [inline templates/http/default.conf.j2]: Maybe we can remove some duplicate code by changing the `ssl` option into a conditional check, moving the `item.value.ssl` conditional check to only wrap around the `ssl` directives and then removing the listen block at https://github.com/nginxinc/ansible-role-nginx/pull/147/files#diff-ec7d84372203d5d7af253681f4c6fc67R78?

Comment (shokinn) [reply templates/http/default.conf.j2]: Sounds good to me.
I'll change my PR.

File: README.md [hunk 2]
```diff
@@ -355,7 +355,11 @@ nginx_http_template:
     template_file: http/default.conf.j2
     conf_file_name: default.conf
     conf_file_location: /etc/nginx/conf.d/
-    port: 8081
+    listen:
+      port: 8081
+      opts: [] # Listen opts like http2 which will be added (ssl is automatically added if you specify 'ssl:').
```
Comment (alessfg) [inline README.md]: There's a slight misalignment in this line

File: defaults/main.yml [hunk 3]
```diff
@@ -167,7 +167,11 @@ nginx_http_template:
     template_file: http/default.conf.j2
     conf_file_name: default.conf
     conf_file_location: /etc/nginx/conf.d/
-    port: 8081
+    listen:
+      port: 8081
+      opts: []
```
Comment (alessfg) [inline defaults/main.yml]: Can you include the comment from the README?

File: README.md (hunk 2 above)
Comment (shokinn) [reply README.md]: Hey @alessfg ,
which kind of misalignment? I don't see it.

Comment (alessfg) [reply README.md]: L359 and L360 -- the port and opts parameters are not equally indented :)

Comment (shokinn) [reply README.md]: I checked at here on GitHub and also on my local files. both are equally intended.
I copied this from my local files:  
```
    listen:
      port: 8081
      opts: [] # Listen opts like http2 which will be added (ssl is automatically added if you specify 'ssl:').
      ipv4: true
      ipv6: false
```

I'm sorry. Can you show me how do you think it should look correct?

If you mean that `port` and `opts` are intended under `listen` this is correct, is the mentioned break of the current behavior how to set the port.

Comment (alessfg) [reply README.md]: Oh yeah you are absolutely right! Sorry for the confusion 😕 GitHub was (still is) displaying the listen block in the README with `ports` and `opts` having different indentations 🤷‍♂

File: defaults/main.yml (hunk 3 above)
Comment (shokinn) [reply defaults/main.yml]: @alessfg done :)

File: README.md (hunk 2 above)
Comment (shokinn) [reply README.md]: OK. I thought I had some problems with my eyes 😅 

*Btw. for me It looks like that they have the same indentation.
Quiet weird.

**Comment by mkryshak** [comment]
@alessfg @shokinn can we extend this to allow multiple 'listen' directives (e.g. one for IPv4 and IPv6)?  There can be an address and port variables for IPv4 and a separate address and port variables for IPv6.  This would allow you to bind to specific IPv4 and IPv6 addresses as opposed to binding to '0.0.0.0' and '::'.

**Comment by shokinn** [comment]
@mkryshak So there are multiple listen directives right now (one for IPv4 and one for IPv6).
Do you want to have these directives dynamically like this?
```yaml
    listen:
      listen_ipv4_127.0.0.1:
        ip: "127.0.0.1"
        port: 8081
        opts: [] # Listen opts like http2 which will be added (ssl is automatically added if you specify 'ssl:').
        ipv4: true
        ipv6: false
      listen_ipv6_192.168.178.42:
        ip: "192.168.178.42"
        port: 8082
        opts: [] # Listen opts like http2 which will be added (ssl is automatically added if you specify 'ssl:').
        ipv4: false
        ipv6: true
      listen_n:
        [...]
```

If the example above is what you have in mind then I would recommend to do it in a sperate PR because in my opinion this does not fit to the scope of current PR.

@alessfg What are you thinking about that?

**Comment by mkryshak** [comment]
@shokinn @alessfg I think the only addition beyond what this PR addresses would be to specify the IP address, similar to what you have proposed, by adding an "ip" field.  I can have a second PR opened to specify the IP address but I do feel it does fit this PR because it is part of making the `listen` directive configurable (e.g. allowing you to specify the bind IP address).

**Comment by alessfg** [comment]
I think what @mkryshak means is that at the moment it's not possible to specify the IP address of listen directives. However, supporting multiple listen directives like in your example @shokinn would be even better. In any case I think any changes related to the `listen` directive fit in this PR.

**Comment by shokinn** [comment]
@alessfg Jupp will implement it this or next week when I have some spare time.

**Comment by alessfg** [comment]
I've given it a try at updating the PR myself. I removed the `ipv4` and `ipv6` parameters to simplify the overall configuration and added a comment indicating users that they need to wrap IPv6 addresses in square brackets.

**Comment by shokinn** [comment]
Well, was on my list for today to implement this 😅

Thanks @alessfg for implementing!

**Pull request merged**